### ISSUE TITLE
Dart2JS crashed when compiling example Angular app

### ISSUE THREAD

**Issue by chalin** [dart-lang/sdk#31889]
Here is a link to the Travis log were dart2js reports a crash:
- https://travis-ci.org/dart-lang/site-webdev/jobs/328324370

Here is an excerpt from the log:

```nocode
[15:42:22] Overriding the upper bound Dart SDK constraint to <=2.0.0-dev.17.0 for the following packages:
angular, angular_ast, angular_compiler, angular_components, angular_forms, angular_test, ansicolor, archive, args, async, barback, bazel_worker, boolean_selector, browser, build, build_barback, build_compilers, build_config, build_runner, built_collection, built_value, charcode, cli_util, code_builder, code_transformers, collection, convert, crypto, csslib, dart_style, dart_to_js_script_rewriter, front_end, glob, html, http, http_multi_server, http_parser, intl, io, isolate, js, json_annotation, kernel, logging, matcher, meta, mime, multi_server_socket, node_preamble, observable, package_config, package_resolver, pageloader, path, perf_api, plugin, pool, protobuf, pub_semver, quiver, quiver_hashcode, sass, sass_builder, scratch_space, shelf, shelf_packages_handler, shelf_static, shelf_web_socket, source_gen, source_map_stack_trace, source_maps, source_span, stack_trace, stream_channel, stream_transform, string_scanner, template_syntax, term_glyph, test, tuple, typed_data, unittest, utf, uuid, watcher, web_socket_channel, webdriver, yaml

To disable this you can set the PUB_ALLOW_PRERELEASE_SDK system environment variable to `false`, or you can silence this message by setting it to `quiet`.

[15:42:22] Warning: You are using these overridden dependencies:
! analyzer 0.31.0-alpha.2
[15:42:22] Got dependencies!
[15:42:22] Precompiling dependencies...
[15:42:22] Loading source assets...
[15:42:22] Loading angular, sass_builder and dart_to_js_script_rewriter transformers...
[15:43:49] Precompiled angular, angular_components, angular_forms, angular_test, code_builder, dart_style, dart_to_js_script_rewriter and intl.
[15:43:49] Precompiling executables...
[15:43:49] Loading source assets...
[15:43:49] Loading angular transformers...
[15:43:53] Precompiled test:test.
[15:44:48] Precompiled angular_test:angular_test.
[15:44:48] completed: pub get {"cwd":"examples/ng/doc/template-syntax"}
[15:44:48] async exec: pub build --web-compiler=dart2js {"cwd":"examples/ng/doc/template-syntax"}
[15:44:48] Loading source assets...
[15:44:49] Loading angular, sass_builder, dart_to_js_script_rewriter and test/pub_serve transformers...
[15:44:55] Building template_syntax...
[15:46:17] [Info from Dart2JS]:
Compiling template_syntax|web/main.dart...
[15:46:21] [Error from Dart2JS on template_syntax|web/main.dart]:
[15:46:21] web/packages/angular/src/core/metadata.dart:353:21:
[15:46:21] Local variable without value.
[15:46:21]           selector: selector,
[15:46:21]                     ^^^^^^^^
[15:46:21] The compiler is broken.

When compiling the above element, the compiler crashed. It is not
possible to tell if this is caused by a problem in your program or
not. Regardless, the compiler should not crash.

The Dart team would greatly appreciate if you would take a moment to
report this problem at http://dartbug.com/new.

Please include the following information:

* the name and version of your operating system,

* the Dart SDK build number (build number could not be determined), and

* the entire message you see here (including the full stack trace
  below as well as the source location above).

[15:46:22] Build error:
[15:46:22] Transform Dart2JS on template_syntax|web/main.dart threw error: Internal Error: Local variable without value.
[15:46:22] package:compiler_unsupported/src/compiler.dart 1242                              CompilerDiagnosticReporter.internalError
[15:46:22] package:compiler_unsupported/src/compile_time_constants.dart 1203                ConstructorEvaluator.visitSend
[15:46:22] package:compiler_unsupported/src/tree/nodes.dart 716                             Send.accept
[15:46:22] package:compiler_unsupported/src/compile_time_constants.dart 421                 CompileTimeConstantEvaluator.evaluateConstant
[15:46:22] package:compiler_unsupported/src/elements/elements.dart 787                      Elements.makeArgumentsList
[15:46:22] package:compiler_unsupported/src/compile_time_constants.dart 872                 CompileTimeConstantEvaluator.evaluateArgumentsToConstructor
[15:46:22] package:compiler_unsupported/src/compile_time_constants.dart 1321                ConstructorEvaluator.evaluateConstructorInitializers
[15:46:22] package:compiler_unsupported/src/compile_time_constants.dart 1375                ConstructorEvaluator.evaluateConstructorFieldValues.<fn>
[15:46:22] 
...
```

Environment:
- Dart VM version: 2.0.0-dev.17.0 (Thu Jan 11 14:47:01 2018 +0100) on "linux_x64"

cc @kwalrath @kevmoo @matanlurey

**Comment by chalin** [comment]
I just compiled the same app using dart2js from 2.0.0-dev.22.0 and the compiler still crashes, though now it seems to be for different reasons:

```nocode
/opt/brew/Cellar/dart/2.0.0-dev.22.0/libexec/lib/async/zone.dart:1380:3:
Internal Error: The compiler crashed when compiling this element.
  R runUnary<R, T>(R f(T arg), T arg) {
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The compiler is broken.
...
/opt/brew/Cellar/dart/2.0.0-dev.22.0/libexec/lib/async/zone.dart:952:3:
Internal Error: The compiler crashed when compiling this element.
  ZoneUnaryCallback<R, T> bindUnaryCallback<R, T>(R f(T arg)) {
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
```

**Comment by sayertindall** [comment]
Did you ever end up figuring it out?

**Comment by sayertindall** [comment]
I am having the same issue.

**Comment by chalin** [comment]
> Did you ever end up figuring it out?

@sayertindall - no, I didn't investigate any further.

**Comment by sigmundch** [comment]
@chalin can you include the full stack trace for your second error?

FWIW - the error you saw with 2.0.0-dev.17.0 was fixed in version 2.0.0-dev.21.0.

**Comment by chalin** [comment]
I've upgraded to 2.0.0-dev.23.0 and you are right that the original error is gone.

As for the other error, I think it is because I was using `--strong` at the time (which I've stopped doing for now).

Closing this since the orignal issue is resolved (e.g. see this [Travis log](https://travis-ci.org/dart-lang/site-webdev/builds/340628682)).
